### ISSUE TITLE
[lc_ctrl] Remote DoS attacks

### ISSUE THREAD

**Issue by gdessouky** [lowRISC/opentitan#17118]
### Description

The life cycle state of an OT device can transition to `SCRAP` from any state, by means of configuring the life cycle controller CSRs via SW or JTAG. While this is not an attack that enables privileged access to assets, it brings the device to a permanently terminal state = DoS.  In an offline discussion with @msfschaffner, we realize that if an attacker manages to gain privileged access (by means of SW vulnerabilities) to the system bus and can write transactions to it, then various remote Denial-of-Services attacks would be possible; transition to `SCRAP` would only be one such example.

Does it make sense to discuss this further and evaluate whether we need to consider such class of attacks and if additional hardening countermeasures to the system bus or otherwise are necessary? Also are these attacks a concern/requirement for certification?

@johannheyszl @moidx @msfschaffner

**Comment by tjaychen** [comment]
@cdgori 

Yes this is certainly true.  I think we allowed software to do life cycle transitions as well for ease of use.
I personally think if this is a concern, we can add software lock bits to disable sw side life cycle transitions.
We can then leave the responsibility of setting these locks to someone like `ROM` or `ROM_ext`. 

Either way though, I don't think this is something we have to fix for M3, but probably something more critical for production silicon (unless we discover other vulnerabilities that somehow make it easier to extract secrets after getting into SCRAP state).

**Comment by msfschaffner** [comment]
Agreed, if we identify this as a risk, we could add CSR lock bits for this.
I am labeling this as P4 / M3 for now - since I think it's something we can take care of later for prod.

**Comment by johannheyszl** [comment]
thx @tjaychen and @msf and adding some thoughts from our lunch discussion w/ @gdessouky @jadephilipoom @ballifatih @bilgiday to it:

- Lock bits that are set at end of ROM seem very effective.  We were unsure whether SCRAP'ing must be available later.
- Agree that we can probably leave this for PS.
- Product risk is that exploits for any SW in M-mode would otherwise allow SCRAP'ing (think wide-scale attack on products / datacenters).
- no certification issues since with physical access included in their threat model, there are always super easy DoS vectors (think hammer).

**Comment by gdessouky** [comment]
thx @tjaychen and @msfschaffner, and @johannheyszl for summary of our discussion!

Few more concerns/questions to add that came up while discussing:

- do we understand correctly that these life cycle transitions are allowed at kernel level? 
- is this programmed to OTP via SW (no physical access) and then the transition occurs successfully when the device boots up again after coming out of reset? I am a bit unclear on the SW (kernel)/OTP controller/life cycle controller interaction that has to occur for this transition to `SCRAP` happen.
- as @johannheyszl points out, do we need to enable `SCRAP`ping or any life cycle transitions beyond ROM?

**Comment by msfschaffner** [comment]
sounds good, thanks for your additional thoughts!

regarding the follow-up questions:

- to initiate a life cycle transition, SW needs to write the appropriate commands to the life cycle controller CSRs, which will then make the necessary updates in OTP automatically (see [programmer's guide](https://docs.opentitan.org/hw/ip/lc_ctrl/doc/#programmers-guide)). you are correct that the new life cycle state only takes effect after a reset / reboot.
- hence, if the lc_ctrl CSR region is not locked via by PMP, any SW running on the processor can write to the life cycle CSRs and initiate such a transition.
- I am not sure regarding the SCRAP'ing requirement... maybe something that @moidx or @alphan can answer. we will probably want to retain the ability to move into SCRAP via the JTAG interface (which requires physical presence), and only lock out the SW side at the appropriate stage (ROM/ROM_EXT?).

**Comment by moidx** [comment]
Having a transition lock bit in the future would be useful. From an inventory management perspective, this is something that could be managed in a code partition running at the same privilege as the `ROM_EXT`.  If we need to expose this as a kernel API to cover any specific certification requirement, we could think about adding a boot service command service by the `ROM_EXT`. This would require an additional reset. The idea here is to always disable entry to `SCRAP` after leaving the `ROM_EXT`.

If there are no certification requirements associated with having a `SCRAP` path in production, then I would recommend only making the `SCRAP` entry available during manufacturing stages and from `RMA` (as originally captured in the life cycle architecture diagram).

nit: I think we should group device types in the memory map to defrag the ePMP configuration. At some point we made some changes to group all crypto devices in a contiguous space, but it may be useful to take a look again to also consider grouping other machine mode critical blocks (e.g. `lc_ctrl`, `alert_handler`, `otp_ctrl`, `flash_ctrl`).

**Comment by gdessouky** [comment]
- Going through the certification and protection profile referenced [1] , it seems we need some sort of `SCRAP` state. It is not explicitly required that we have a path from every state to `SCRAP` but for the sake of "being on the safe side" given this ambiguity and to accommodate certain settings and their requirements, I think it probably make sense to keep `SCRAP` accessible from every state including `PROD_END`. 

  - If we were to allow transitions to `SCRAP` only from `RMA`, `PROD_END` would have no path to `SCRAP`, since PROD_END has no path to `RMA`. 

  - Going through the certification requirements, it is also not clear if we need 2 states `PROD` and `PROD_END` but the distinction is clear and might be required in some settings. 


- Locking the transitions after leaving `ROM_EXT` is as good as disabling SW-initiated transitions entirely AFAIU. Currently, the software requests the life cycle transition by writing to the life cyle controller CSRs and this transition actually takes effect after a reset anyway. Locking the CSRs at the end of `ROM_EXT` means no SW can request this transition altogether, which is as good as fundamentally disabling SW transitions entirely. Does that make sense or am I missing something?


- Evolving it into a boot service command (in `ROM_EXT` like attestation and ownership transfer) does not seem to help mitigate the DoS issue either, according to my understanding. Again, software can request this boot service by placing the request structure in retention SRAM and the transition would take effect after reboot, so any software can still request the life cycle transition, right? I am not entirely sure how are the mechanics for invoking these boot services intended and if they will be authenticated somehow? Maybe you can help me understand this better.


Given the above (and also the current implementation), I would propose the following:
- `lc_ctrl.TRANSITION_REGWEN` is currently set by default to 0 and is only flipped by the hardware to 1 when the mutex is claimed by software. Only then is a SW-initiated life cycle state transition enabled. We can add another `SW-TRANSITION_EN` bit that is set by default to 0 and it locks flipping of `lc_ctrl.TRANSITION_REGWEN` to 1, i.e., disables SW-initiated transfers by default.
- At ownership transfer in ROM_EXT, the owner selects if SW-initiated transitions would stay by default locked or be enabled (at owner's own peril). If enabled, that new `SW-TRANSITION_EN` bit is hardware-flipped to 1. It can only be flipped one more time (to the default 0) if an owner requires the additional measure later during its ownership lifetime, but never back again to 1 except at the next ownership transfer.

So in short, for now, until ownership transfer is actually implemented in ROM_EXT, SW-initiated transitions would be effectively disabled.

Let me know your thoughts and if this makes sense @moidx @johannheyszl  @cindychip and please tag anyone else who can pitch in.

[1] https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Zertifizierung/Reporte/ReportePP/pp0084b_pdf.pdf?__blob=publicationFile&v=1

**Comment by gdessouky** [comment]
Triaged for `IP:lc_ctrl` - if exploited, this can cause permanent DoS remotely (via SW), i.e., implications are irreversible so ought to be fixed. Pending follow-up to converge on a fix plan.

**Comment by gdessouky** [comment]
Summarizing our discussions in Security WG and follow-ups:
- Safest not to allow to SCRAP (at least via SW) from PROD or DEV states
- But logical scrapping (and also remotely) might be required for certain use cases, e.g., secure element, so long as it is triggered from actors with the relevant privileges
- To allow compliance for the different use cases and to keep changes minimal for M2.5, we agree to introduce a lock bit to the life cycle CSRs that gets locked in ROM_EXT if required (per SKU)
- For future releases, introducing an authenticated boot service to enable the transition and/or removing the capability of SW transitions in PROD and DEV life cycle states altogether (if not required by certification for the target use case) would be considered
- Other routes to remote DoS attacks could be possible and worth looking into, e.g., countermeasures overreacting and locking up the chip, or using an RMA bad token too many times and permanently terminating the device, etc.

**Comment by johngt** [comment]
@moidx / @gdessouky - assuming if backlog we can put as P3 for now? If incorrect please change back to P2 or lower.

**Comment by johannheyszl** [comment]
@gdessouky is already working on adding the lock bit for M2.5. Labeling as P1 again since will be done soon. Is not on backlog atm, which makes sense.

After this, @gdessouky pls to state SW follow-up requirements and other future work and we will relabel for this then.
